rpcs3/Emu/system_progress.hpp

```diff
@@ -15,21 +27,52 @@ extern atomic_t<bool> g_system_progress_stopping;
 class scoped_progress_dialog final
 {
 	// Saved previous value
-	const char* const m_prev;
+	const char* m_prev;
+	u32 m_prev_update;
 
 public:
 	scoped_progress_dialog(const char* text) noexcept
-		: m_prev(g_progr.exchange(text ? text : ""))
 	{
+		std::tie(m_prev, m_prev_update) = g_progr.atomic_op([text = text ? text : ""](progress_dialog_string_t& progr)
+		{
+			const char* old = progr.m_text;
+			progr.m_user_count++;
+			progr.m_updates++;
+			progr.m_text = text;
+			return std::make_pair(old, progr.m_updates);
+		});
 	}
 
 	scoped_progress_dialog(const scoped_progress_dialog&) = delete;
 
 	scoped_progress_dialog& operator=(const scoped_progress_dialog&) = delete;
 
+	scoped_progress_dialog& operator=(const char* text) noexcept
+	{
+		g_progr.atomic_op([text = text ? text : ""](progress_dialog_string_t& progr)
+		{
+			progr.m_text = text;
+			progr.m_updates++;
+		});
+
+		return *this;
+	}
+
 	~scoped_progress_dialog() noexcept
 	{
-		g_progr.release(m_prev);
+		g_progr.atomic_op([this](progress_dialog_string_t& progr)
```

Comment:
I don't see the purpose in this.
You're supposed to reset g_progr on scope exit.
It's not supposed to be updated by 2 different entities simultaneously.

Reply:
Well, it kind of is updated by 2 different PPUs if they call the same syscalls sys_prx or sys_overlay syscalls.
The perfect mechanism would be keeping a thread-safe array of strings but it seems overkill for this purpose.